File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Unit/Monster/MonsterGroup.ts
import { Mathf } from "../../../../../Core/Module/Math/Mathf";
import DungeonManager from "../../DungeonManager";
import { AbstractMonster } from "./AbstractMonster";
import { Intent } from "./Intent";

export class MonsterGroup{
    public Monsters = new Array<AbstractMonster>();
    public constructor(...monsters:Array<AbstractMonster>) {
        this.Monsters = monsters;
    }

    public AreMonstersBasicallyDead() : boolean {
        return this.GetAliveMonstersCount() == 0;
    }
    
    public GetAliveMonstersCount():number{
        let aliveCount = 0;
        for (let i = 0; i < this.Monsters.length; i++) {
            let m = this.Monsters[i];
            if(m.IsDying || m.IsEscaping) continue;
            aliveCount++;
        }
        return aliveCount;
    }
    
    public Random(){
        let vaildMonster = this.Monsters.filter(t1=>!t1.IsDying && t1.Intent != Intent.ESCAPE && !t1.IsDeadOrEscaped)
        let random = Mathf.Floor(Mathf.RandomRange(0,vaildMonster.length,DungeonManager.Inst.CurrentDungeon.AIRng));
        return vaildMonster[random];
    }
    
    public IndexOf(m : AbstractMonster) : number{
        return this.Monsters.indexOf(m);
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Unit/Monster/MonsterGroup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using ModTheSpire.Dungeon;
using UnityEngine;

namespace ModTheSpire.Unit.Monster
{
    public class MonsterGroup{
        public List<AbstractMonster> Monsters;
        public MonsterGroup(List<AbstractMonster> monsters) {
            this.Monsters = monsters;
        }

        public bool AreMonstersBasicallyDead() {
            return GetAliveMonstersCount() == 0;
        }
    
        public int GetAliveMonstersCount(){
            var aliveCount = 0;
            for (var i = 0; i < this.Monsters.Count; i++) {
                var m = this.Monsters[i];
                if(m.IsDying || m.IsEscaping) continue;
                aliveCount++;
            }
            return aliveCount;
        }
    
        public AbstractMonster Random()
        {
            var vaildMonster = this.Monsters.Where(t1 => !t1.IsDying && t1.Intent != Intent.ESCAPE && !t1.IsDeadOrEscaped).ToList();
            var random = Mathf.FloorToInt(DungeonManager.Inst.CurrentDungeon.AIRng.Next(0,vaildMonster.Count));
            return vaildMonster[random];
        }
    
        public int IndexOf(AbstractMonster m)
        {
            return Monsters.FindIndex(t1 => t1 == m);
        }

        public void AddMonster(AbstractMonster m)
        {
            this.Monsters.Add(m);
        }
    }
}